src/Interpreters/Aggregator.cpp

```diff
@@ -855,12 +855,18 @@ void NO_INLINE Aggregator::executeWithoutKeyImpl(
 
 
 void NO_INLINE Aggregator::executeOnIntervalWithoutKeyImpl(
-    AggregatedDataWithoutKey & res,
+    AggregatedDataVariants & data_variants,
     size_t row_begin,
     size_t row_end,
     AggregateFunctionInstruction * aggregate_instructions,
-    Arena * arena)
+    Arena * arena) const
 {
+    /// `data_variants` will destroy the states of aggregate functions in the destructor
+    data_variants.aggregator = this;
```

Comment:
@azat do not understand why this is required, please provide more information.

Reply:
Like comment stated, it is required to destroy states, see https://github.com/ClickHouse/ClickHouse/blob/00330461d1eefc541cff813fa5b8981cd4585b27/src/Interpreters/Aggregator.cpp#L52-L65

Also note, that usually, there is no leak since if `Aggregator::addSingleKeyToAggregateColumns` finished then states from `data_variants` will be moved to `aggregate_columns` and cleaned properly.

Reply:
Yes that make sense.
Another question why do we need to call `init`
```c++
data_variants.init(AggregatedDataVariants::Type::without_key);
```
It seems that for without_key it will not create any state, but just change AggregatedDataVariants type, as I understand we did not change it to without_key, in AggregatingInOrderTransform ?

Reply:
>Another question why do we need to call init

Otherwise destroy is no-op

https://github.com/ClickHouse/ClickHouse/blob/437940b29d2e8e95d214ecfbdfcb66b52c7b2cdd/src/Interpreters/Aggregator.cpp#L2775-L2778

https://github.com/ClickHouse/ClickHouse/blob/437940b29d2e8e95d214ecfbdfcb66b52c7b2cdd/src/Interpreters/Aggregator.h#L636

Reply:
Yes, make sense, thanks for explanation.

Reply:
Yeah, the code is pretty tricky, and w/o worked LSan right now (#33957 had been partially reverted in #34146) it is pretty easy to introduce new leaks...